Clean up schema documentation, standardize units

rly opened the pull request:
Fixes https://github.com/NeurodataWithoutBorders/nwb-schema/issues/279 and https://github.com/NeurodataWithoutBorders/nwb-schema/issues/277

`doc` attributes were modified. In some cases, `unit` was modified to be consistent. In some cases, order of attributes were shuffled to put most important ones first (so that documentation is easier to read).

rly committed:
Reorder keys in YAML files, add dims next to shape, add line spacing

rly committed:
Merge branch 'dev' into enh/reorg_yaml

rly committed:
Merge branch 'dev' into enh/reorg_yaml

rly committed:
Fix quotes missing around quantity: '+'

rly committed:
Updated docstrings for base, behavior, ecephys, epoch, and file

rly committed:
Update more docstrings

rly committed:
Update misc and ogen docstrings

rly committed:
Update ophys and retinotopy docs

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
Update docs

rly committed:
Make all units consistent

See https://github.com/NeurodataWithoutBorders/nwb-schema/issues/277

bendichter commented on src/pynwb/data/nwb.base.yaml:
```diff
@@ -18,29 +18,33 @@ datasets:
 - neurodata_type_def: VectorData
   neurodata_type_inc: NWBData
-  doc: Data values indexed by pointer
+  doc: A dataset indexed by pointer (see VectorIndex). Together with VectorIndex,
```
`VectorData` is also used for `DynamicTable` columns that are not indexed.

bendichter commented on src/pynwb/data/nwb.behavior.yaml:
```diff
@@ -28,15 +28,20 @@ groups:
-        \ below describes how to convert the data to the specified SI unit."
+      doc: "Base unit of measurement for working with the data. The default value
+        is 'meter'. If the data are stored in acquisition system units or other units
+        that require a conversion to be interpretable, the attribute 'conversion'
```
This seems more like a description of 'conversion' than of 'unit'.

bendichter commented on src/pynwb/data/nwb.ecephys.yaml:
```diff
@@ -32,24 +32,29 @@ groups:
-    \ [num channels] [num samples] (or [num events] [num samples] for single electrode)."
+  doc: "Stores snapshots/snippets of recorded spike events (i.e., threshold crossings). This
+    may also be raw data, as reported by ephys hardware. If so, the TimeSeries::description
+    field should describing how events were detected. All SpikeEventSeries should
```
```suggestion
    field should describe how events were detected. All SpikeEventSeries should
```

bendichter commented on src/pynwb/data/nwb.ecephys.yaml:
```diff
@@ -328,25 +251,113 @@ groups:
   - name: location
     dtype: text
-    doc: description of location of this electrode group
+    doc: Location of this electrode group.
```
We should give advice here for specific ontologies that we recommend for naming anatomical regions.

bendichter commented on src/pynwb/data/nwb.ecephys.yaml:
```diff
@@ -248,28 +171,28 @@ groups:
       dtype: text
-      default_value: Seconds
-      doc: The string ''Seconds''
+      default_value: second
```
```suggestion
      value: second
```

bendichter commented on src/pynwb/data/nwb.file.yaml:
```diff
@@ -140,78 +141,76 @@ groups:
       dtype: text
-      doc: 'Name of person who performed the experiment.COMMENT: More than one person
-        OK. Can specify roles of different people involved.'
+      doc: Name of person(s) who performed the experiment. Can also specify roles
```
We might want to change this to be a list in order to list multiple experimenters

rly replied:
Agreed, but that is a separate issue. I did not want to pack too many changes into this PR.

bendichter commented on src/pynwb/data/nwb.icephys.yaml:
```diff
@@ -94,100 +93,78 @@ groups:
-        \ unit. COMMENT: This is the SI unit (when appropriate) of the stored data,\
-        \ such as Volts. If the actual data is stored in millivolts, the field 'conversion'\
-        \ below describes how to convert the data to the specified SI unit."
+      default_value: farad
```
```suggestion
      value: farad
```

bendichter commented on src/pynwb/data/nwb.icephys.yaml:
```diff
@@ -94,100 +93,78 @@ groups:
-        \ unit. COMMENT: This is the SI unit (when appropriate) of the stored data,\
-        \ such as Volts. If the actual data is stored in millivolts, the field 'conversion'\
-        \ below describes how to convert the data to the specified SI unit."
+      default_value: farad
```
```suggestion
      value: farad
```

bendichter commented on src/pynwb/data/nwb.icephys.yaml:
```diff
@@ -94,100 +93,78 @@ groups:
-        \ unit. COMMENT: This is the SI unit (when appropriate) of the stored data,\
-        \ such as Volts. If the actual data is stored in millivolts, the field 'conversion'\
-        \ below describes how to convert the data to the specified SI unit."
+      default_value: ohm
```
```suggestion
      value: ohm
```

bendichter commented on src/pynwb/data/nwb.image.yaml:
```diff
@@ -81,10 +80,10 @@ groups:
+      This is only relevant if the image is stored in the file system as one
       or more image file(s). This field should NOT be used if the image is stored
-      in another HDF5 file and that file is HDF5 linked to this file.'
+      in another NWB file and that file is HDF5 linked to this file.
```
Maybe we shouldn't mention HDF5 in the schema, since this is supposed to back independent of backend

bendichter commented on src/pynwb/data/nwb.misc.yaml:
```diff
@@ -76,11 +74,13 @@ groups:
+      doc: Base unit of measurement for working with the data. Annotations have
+        no units, so the value is fixed to 'n/a'.
 
 - neurodata_type_def: IntervalSeries
```
Do we have a purpose for this? I feel like the Intervals table can handle this type of data.

rly replied:
I agree. This was in the 1.0.6 spec, so I gather it is a legacy item that can now be deprecated. @oruebel @ajtritt what do you think?

ajtritt replied:
@rly yes it got carried along from the beginning. I'm not aware of anyone using it now though.

bendichter commented on src/pynwb/data/nwb.ecephys.yaml:
```diff
@@ -328,25 +251,113 @@ groups:
+  links:
+  - name: clustering_interface
+    target_type: Clustering
+    doc: HDF5 link to Clustering interface that was the source of the clustered data
```
```suggestion
    doc: Link to Clustering interface that was the source of the clustered data
```

bendichter reviewed (commented):
Thanks for cleaning this up! I see two remaining issues we should decide on:
1) Do we want to mention HDF5 in the schema?
2) Do we want to use singular or plural for SI units?

rly committed:
Remove references to HDF5

rly committed:
Make some default units fixed:
ecephys:
- TimeSeries.data 'unit' attribute is fixed to 'volt'
- SpikeEventSeries.data 'unit' attribute is fixed to 'volt'
- EventDetection.times 'unit' attribute is fixed to 'second'

icephys:
- VoltageClampSeries.capacitance_fast 'unit' attribute is fixed to 
'farad'
- VoltageClampSeries.capacitance_slow 'unit' attribute is fixed to 
'farad'
- VoltageClampSeries.resistance_comp_bandwidth 'unit' attribute is fixed 
to 'hertz'
- VoltageClampSeries.resistance_comp_correction 'unit' attribute is 
fixed to 'percent'
- VoltageClampSeries.resistance_comp_prediction 'unit' attribute is 
fixed to 'percent'
- VoltageClampSeries.whole_cell_capacitance_comp 'unit' attribute is 
fixed to 'farad'
- VoltageClampSeries.whole_cell_series_resistance_comp 'unit' attribute 
is fixed to 'ohm'

ogen:
- OptogeneticSeries.data 'unit' attribute is fixed to 'watt'

rly committed:
Clean up docstrings

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
Update src/pynwb/data/nwb.ecephys.yaml

Co-Authored-By: Ben Dichter <ben.dichter@gmail.com>

rly committed:
Update SpatialSeries units

rly committed:
Merge branch 'enh/yaml_docs' of https://github.com/NeurodataWithoutBorders/pynwb into enh/yaml_docs

rly committed:
Clean up 'location' attribute doc

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
Pluralize all units in schema

rly committed:
Update tests to use right units

rly committed:
Update class init to use right units

rly committed:
Add test to test linking of electricalseries (check shape)

Requires hdmf PR #104

rly committed:
Remove tests of container parent, add tests of nwbfields

rly committed:
Add check for ElectricalSeries data shape, enhance linking test

rly committed:
Refactor docval usage, set some defaults to match spec better

rly committed:
Put the timeseries tests in the same place

rly committed:
flake8 fixes on tests

rly committed:
Merge branch 'enh/yaml_docs' of https://github.com/NeurodataWithoutBorders/pynwb into enh/yaml_docs

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly commented:
This also resolves #1013 and #1005

rly committed:
Set value of datasets in IZeroClampSeries spec to 0

rly committed:
Make unit arg to IndexSeries no longer required to match spec

rly committed:
Make unit arg to ImageMaskSeries no longer required to match spec

rly committed:
Make unit arg to OpticalSeries no longer required to match spec

rly committed:
Remove hardcoding of 'conversion'.

It is set by default in TimeSeries

rly committed:
Specify default unit as "no unit" for DecompositionSeries

rly committed:
Make description arg to DecompSeries no longer required to match spec

rly committed:
Make unit arg to RoiResponseSeries no longer required to match spec

rly committed:
Bug fix on change to DecompositionSeries spec

rly committed:
Clean up spec changes for IZeroClampSeries

rly commented:
### PyNWB changes:
- TimeSeries.unit: fixed value "Seconds" -> fixed value "seconds"
- ElectricalSeries.unit: fixed value "volt" -> fixed value "volts"
- OptogeneticSeries.unit: default value "Watt" -> fixed value "watts"

- ImageSeries.unit: default value "None" -> default value None

- TimeSeries.timestamps: shape unspecified -> (None,) to match spec

- AnnotationSeries.data: shape unspecified -> (None,) to match spec

- IntervalSeries.data: default value list() -> None to match all other TimeSeries

- Remove ability to set `parent` or `container_source` in constructor of all types (see also https://github.com/hdmf-dev/hdmf/pull/90 which removes these arguments from Container)
- The bias_current, bridge_balance, and capacitance_compensation arguments of IZeroClampSeries can no longer be set. Their values are fixed to 0. Previously they were settable and had default value of 0 -- see #1013

- ImageMaskSeries.dimension: default value [np.nan] -> default value None (spec does not specify it)
- OpticalSeries.dimension: default value [np.nan] -> default value None (spec does not specify it)
- TwoPhotonSeries.dimension: default value [np.nan] -> default value None (spec does not specify it)

#### The following are API breaking changes because the order of constructor arguments changes:

- ImageMaskSeries.format: required -> not required (spec does not require it)
- OpticalSeries.format: required -> not required (spec does not require it)

- IndexSeries.unit: required -> not required (spec does not require it)
- ImageMaskSeries.unit: required -> not required (spec does not require it)
- OpticalSeries.unit: required -> not required (spec does not require it)
- RoiResponseSeries.unit: required -> not required (spec does not require it)

- DecompositionSeries.description: required -> not required (spec does not require it)

This is not backwards compatible because optional arguments are parsed after required arguments, so code originally written like this:
`ts = RoiResponseSeries('test_ts', list(), 'unit', rt_region, timestamps=list())`
where the first four arguments (name, data, unit, rois) were required, would now break because docval would set rois to 'unit'.

### NWB Schema changes:
- TimeSeries dataset "starting_time" attribute "unit" fixed value: "Seconds" -> "seconds"
- TimeSeries dataset "timestamps" attribute "unit" fixed value: "Seconds" -> "seconds"
- ElectricalSeries dataset "data" attribute "unit" default value: "volt" -> fixed value "volts"
- SpikeEventSeries dataset "data" attribute "unit" default value: "volt" -> fixed value "volts"
- SpikeEventSeries dataset "timestamps" attribute "unit" fixed value: "Seconds" -> fixed value "seconds"
- EventDetection dataset "times" attribute "unit" default value: "Seconds" -> "seconds"
- VoltageClampSeries dataset "capacitance_fast" attribute "unit" default value "Farad" -> fixed value "farads"
- VoltageClampSeries dataset "capacitance_slow" attribute "unit" default value "Farad" -> fixed value "farads"
- VoltageClampSeries dataset "resistance_comp_bandwidth" attribute "unit" default value "Hz" -> fixed value "hertz"
- VoltageClampSeries dataset "resistance_comp_correction" attribute "unit" default value "percent" -> fixed value "percent"
- VoltageClampSeries dataset "resistance_comp_prediction" attribute "unit" default value "percent" -> fixed value "percent"
- VoltageClampSeries dataset "whole_cell_capacitance_comp" attribute "unit" default value "Farad" -> fixed value "farads"
- VoltageClampSeries dataset "whole_cell_series_resistance_comp" attribute "unit" default value "Ohm" -> fixed value "ohms"
- IZeroClampSeries dataset "bias_current" unspecified value -> fixed value 0.0
- IZeroClampSeries dataset "bridge_balance" unspecified value -> fixed value 0.0
- IZeroClampSeries dataset "capacitance_compensation" unspecified value -> fixed value 0.0
- OptogeneticSeries dataset "data" attribute "unit" default value "watt" -> fixed value "watts"
- ImagingPlane dataset "manifold" attribute "unit" default value "Meter" -> default value "meters"
- DecompositionSeries dataset "data" attribute "unit" default value unspecified -> default value "no unit"

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly commented:
@bendichter @oruebel please let me know your thoughts on the API breaking changes I noted above. In some sense, this is fixing a bug -- a discrepancy between the API and the spec. It would be nice to make this change with a gradual warning period, but with docval the way it is, I'm not sure how to do that.

rly commented:
@ajtritt your feedback would also be appreciated

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
Fix tests where unit was required

rly committed:
Merge branch 'dev' into enh/yaml_docs

ajtritt committed:
Merge branch 'dev' into enh/yaml_docs

ajtritt commented on tests/unit/test_timeseries.py:
```diff
@@ -1,50 +0,0 @@
-import unittest
```
Is this file empty now, or am I just confused by Github's rendering?

rly replied:
I deleted this file and moved the tests into tests/unit/test_base.py next to the other TimeSeries tests.

rly committed:
Merge branch 'enh/yaml_docs' of https://github.com/NeurodataWithoutBorders/pynwb into enh/yaml_docs

rly committed:
fix typo

rly committed:
Use hdmf==1.1.2

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
Add missing import

rly committed:
Fix gallery example

rly committed:
Fix missing output of gallery tests

rly committed:
Undo test.py change

rly committed:
Test missing test on python35 in azure

rly committed:
Merge branch 'dev' into enh/yaml_docs

rly committed:
prevent double triggering of setupcontainer in test

rly committed:
Remove percent symbol from docval

rly committed:
temporarily disable verbose on tox

rly committed:
Remove unit attribute of VoltageClampSeries parameters

rly committed:
Revert tox.ini change

rly committed:
Make subtest failures/errors print in verbose mode

rly committed:
Remove unit attribute of remaining VoltageClampSeries parameter

rly committed:
Revert "Remove unit attribute of VoltageClampSeries parameters"

This reverts commit 0983e005cbf21a67a3d7e710e3992912cdbd36f6.

rly committed:
Revert "Remove unit attribute of remaining VoltageClampSeries parameter"

This reverts commit d661659b6fe7bf374fe013dc358c324e122f4715.

rly committed:
Explicitly map all remapped values

rly commented:
This is finally ready to merge.
